x/lazyledgerapp/types/payformessage.go

```diff
@@ -0,0 +1,294 @@
+package types
+
+import (
+	"crypto/sha256"
+	"errors"
+	fmt "fmt"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lazyledger/lazyledger-core/crypto/merkle"
+	core "github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/nmt"
+)
+
+const (
+	TypeMsgPayforMessage                   = "payformessage"
+	TypeSignedTransactionDataPayForMessage = "signedtransactiondatapayformessage"
+	ShareSize                              = core.ShareSize
+	SquareSize                             = 64
+	NamespaceIDSize                        = core.NamespaceSize
+)
+
+///////////////////////////////////////
+// 	MsgWirePayForMessage
+///////////////////////////////////////
+
+var _ sdk.MsgRequest = &MsgWirePayForMessage{}
+
+func (msg *MsgWirePayForMessage) Route() string { return RouterKey }
+
+func (msg *MsgWirePayForMessage) Type() string { return TypeMsgPayforMessage }
+
+// ValidateBasic checks for valid namespace length, declared message size, share
+// commitments, signatures for those share commitments, and fulfills the sdk.Msg
+// interface
+func (msg *MsgWirePayForMessage) ValidateBasic() error {
+	pubK := msg.PubKey()
+
+	// ensure that the namespace id is of length == 8
+	if len(msg.GetMessageNameSpaceId()) != NamespaceIDSize {
+		return fmt.Errorf(
+			"invalid namespace length: got %d wanted 8",
+			len(msg.GetMessageNameSpaceId()),
+		)
+	}
+
+	// ensure that the included message is evenly divisble into shares
+	if uint64(len(msg.GetMessage()))%ShareSize != 0 {
+		return fmt.Errorf("Share message must be divisible by %d", ShareSize)
+	}
```

Comment:
The message size in bytes may not be divisible by `ShareSize`. The consensus rules spec actually uses `ciel` on the non-integer division https://github.com/lazyledger/lazyledger-specs/blob/8977c91d0ceb01b375df1488eb7f08fdb4e4f006/specs/consensus.md#signedtransactiondatapayformessage. Bytes up to a multiple of `ShareSize` are padded with zeroes.

We're free to change this in the spec, but currently the spec and this check don't match.

Reply:
I *think* this is resolved in https://github.com/lazyledger/lazyledger-app/pull/21/commits/96ffd13f3308b03380317985e5b9bfcba574c8d9